### ISSUE TITLE
Add Gerchberg Saxton as a phase retrieval approach

### ISSUE THREAD

**Issue by hosseybposh** [chromatix-team/chromatix#139]
We can add it to CGH tutorial. 2D is simple but for 3D there are different approaches to implement it. We can stick to 2D for now

**Comment by diptodip** [comment]
Do you want to do it if you have time? :)
If not I'll get to it later. I definitely want to add these kinds of algorithms.

**Comment by hosseybposh** [comment]
yup I can implement it. I'll get to it this week I need it for CGH

**Comment by CAClaveau** [comment]
Hi, Thank you for sharing this great tool - it looks fantastic! I was just wondering if there have been any updates on the development and addition of this new feature/algorithm for phase retrieval (linked to [issue #122](https://github.com/chromatix-team/chromatix/issues/122)).

**Comment by diptodip** [comment]
By the way, I'd be happy to help explain these examples more if needed!

**Comment by CAClaveau** [comment]
Hi @diptodip,

I'm beginning to familiarize myself with Chromatix and experimenting with the shared code - I really like it. The documentation and examples are very helpful. I would have a few questions and comments:

1. What is the best strategy to model/simulate a field coming from an off-axis object point source can be arbitrarily placed within the optical system’s FOV? I’m referring to something like the function `objective_point_source` but without applying a lens transformation. (Alternatively, one option would be to first use an FFLens illuminated by an inclined plane wave with a carefully chosen `kykx` so that the output is a point source image with the desired offset?)

2.  The system I would like to simulate might be too large, which could require resampling of the field during propagation. The  functions `transform_propagate` handle resampling well - I assume this is why they are not vectorized with respect to the distance parameter `z`? However, these functions do not support off-axis propagation, they don't have an orientation parameter “kykx”. In contrast, the functions “asm_propagate” and “exact_propagate” support off-axis propagation but do not perform field resampling. Is that correct? Is there a recommended strategy in this case when the maximum sampling is limited, resulting in aliasing, and ultimately erroneous output? I noticed mentions of chirp-z, Bluestein, and matrix Fourier transforms in the repo, which might help achieve smarter frequency sampling. (Also, I think the work here [link](https://whywww.github.io/LSASM_page/) introduces a new off-axis propagation method with resampling. A PyTorch version of their algorithm is available on GitHub.)

3. On a secondary note, when I jit-compile the function `forward` and then try to access the returned field’s attribute `intensity`, I get the following error:
 "`UnexpectedTracerError: Encountered an unexpected tracer. A function transformed by JAX had a side effect, allowing for a reference to an intermediate value with type float64[1] wrapped in a DynamicJaxprTracer to escape the scope of the transformation. JAX transformations require that functions explicitly return their outputs, and disallow saving intermediate values to global state.`"
The same issue occurs with the attribute `field.spectral_density` (which I think is used internally to compute “intensity”). I can understand that such errors may arise in JAX due to its pure functional paradigm, and that one way to avoid this is to return `field.intensity` directly as output from the function `forward`. However, I’m surprised that other attributes like `field.phase` or `field.amplitude` do not trigger the same error.

Thank you for your time!

**Comment by diptodip** [comment]
Thanks for the kind words! It's really cool to hear you're getting some use out of the library.

1. I think this is just an omission on our part. The normal `point_source` could certainly also have an offset term. Though are you also interested in simulating multiple point sources at once?
2. Yes, `transform_propagate` and `transform_propagate_sas` both change the pixel size (and therefore are not automatically batched like `asm_propagate` or `exact_propagate` --- right now a `Field` can vary in pixel size only across the wavelength dimension, unless you use `vmap`). If you don't actually need to propagate off axis but to propagate a large distance, `transform_propagate_sas` ([scalable angular spectrum](https://opg.optica.org/optica/fulltext.cfm?uri=optica-10-11-1407&id=541154)) should do the trick. Though I should mention that the note about negative propagation in the docs for `transform_propagate_sas` is not totally true right now, you don't get back the original field if you propagate a negative distance. But yes as you've noticed @ebezzam is working on a CZT implementation (PR #150), which will help us implement shifted ASM as well. And we should also be able to add LS-ASM. In short, a lot of these features are coming soon!
3. Sorry about that, it's not a very intuitive part of the library right now. When you get attributes like `field.phase` or `field.amplitude`, those are actually just functions of `field.u`, which to JAX is a PyTree node that will not end up as a traced value. But when you call something like `field.spectral_density` or `field.intensity`, those are functions of an internal value to the `Field` which is not a PyTree node, for various reasons. We might be able to change that in the future, but for now the easiest solution is to return multiple values with all the attributes you need from your `forward`.

**Comment by CAClaveau** [comment]
Hi @diptodip ,

Thank you again for your quick reply and detailed explanations!

1- Understood, this is not a big deal for me. I can also refer to the function `objective_point_source` to achieve what I need. At this stage, I don’t need to simulate multiple point sources at once.

2- Thanks for the clarifications and comments on this. Yes, I can use `vmap` as you suggested to vectorize these functions with respect to the distance parameter `z`. The "Parallelism" section in the doc is also very helpful in this regard. Yes, I'll experiment with `transform_propagate_sas`. Looking forward to the upcoming improved propagators!

3- Very clear. I was mainly curious about what was happening with these attributes. We have a simple workaround to bypass the issue.
